client/client.go

```diff
@@ -0,0 +1,109 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/go-http-utils/headers"
+	"go.opencensus.io/plugin/ochttp"
+
+	"github.com/SKF/go-rest-utility/client/auth"
+)
+
+type Doer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type Client struct {
+	BaseURL       *url.URL
+	TokenProvider auth.TokenProvider
+
+	client Doer
+}
+
+func NewClient(opts ...Option) *Client {
+	client := &Client{
+		BaseURL:       nil,
+		TokenProvider: nil,
+		client: &http.Client{
+			Transport: &ochttp.Transport{},
+		},
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client
+}
+
+func (c *Client) Do(ctx context.Context, r *Request) (*Response, error) {
+	httpRequest, err := c.prepareRequest(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+
+	httpResponse, err := c.client.Do(httpRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.prepareResponse(httpResponse)
+}
+
+func (c *Client) prepareRequest(ctx context.Context, req *Request) (*http.Request, error) {
+	url, err := req.toURL(c.BaseURL)
+	if err != nil {
+		return nil, err
+	}
+
+	httpRequest, err := http.NewRequestWithContext(ctx, req.method, url.String(), req.body)
+	if err != nil {
+		return nil, err
+	}
+
+	if req.header.Get(headers.UserAgent) == "" {
+		req.header.Set(headers.UserAgent, "go-rest-utility/v1")
+	}
+
+	if c.TokenProvider != nil {
+		token, err := c.TokenProvider.GetIdentityToken(ctx)
+		if err != nil {
+			return nil, err
```

Comment:
It is always nice extend the errors with fmt.errorf..
This is how we build stack traces in go, if I've understodd i corectly

Reply:
Think I wrapped all relevant ones